File: app/controllers/TimeController.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers;

import be.objectify.deadbolt.java.actions.Group;
import be.objectify.deadbolt.java.actions.Restrict;
import controllers.base.BaseController;
import io.ebean.DB;
import java.io.IOException;
import javax.inject.Inject;
import models.Exam;
import models.ExamEnrolment;
import models.User;
import org.joda.time.DateTime;
import org.joda.time.Seconds;
import play.mvc.Http;
import play.mvc.Result;
import sanitizers.Attrs;
import security.Authenticated;
import util.datetime.DateTimeHandler;

public class TimeController extends BaseController {

    private final DateTimeHandler dateTimeHandler;

    @Inject
    public TimeController(DateTimeHandler dateTimeHandler) {
        this.dateTimeHandler = dateTimeHandler;
    }

    @Authenticated
    @Restrict({ @Group("STUDENT") })
    public Result getRemainingExamTime(String hash, Http.Request request) throws IOException {
        User user = request.attrs().get(Attrs.AUTHENTICATED_USER);
        ExamEnrolment enrolment = DB
            .find(ExamEnrolment.class)
            .fetch("externalExam")
            .where()
            .disjunction()
            .eq("exam.hash", hash)
            .eq("externalExam.hash", hash)
            .endJunction()
            .eq("user.id", user.getId())
            .findOne();

        if (enrolment == null) {
            return notFound();
        }

        DateTime reservationStart = getStart(enrolment);
        int durationMinutes = getDuration(enrolment);
        DateTime now = getNow(enrolment);
        Seconds timeLeft = Seconds.secondsBetween(now, reservationStart.plusMinutes(durationMinutes));

        return ok(String.valueOf(timeLeft.getSeconds()));
    }

    private DateTime getStart(ExamEnrolment enrolment) {
        if (enrolment.getExaminationEventConfiguration() != null) {
            return enrolment.getExaminationEventConfiguration().getExaminationEvent().getStart();
        }
        return enrolment.getReservation().getStartAt();
    }

    private DateTime getNow(ExamEnrolment enrolment) {
        if (enrolment.getExaminationEventConfiguration() != null) {
            return DateTime.now();
        }
        return dateTimeHandler.adjustDST(DateTime.now(), enrolment.getReservation());
    }

    private int getDuration(ExamEnrolment enrolment) throws IOException {
        if (enrolment.getExam() != null) {
            return enrolment.getExam().getDuration();
        }
        Exam exam = enrolment.getExternalExam().deserialize();
        return exam.getDuration();
    }
}

File: app/controllers/examination/TimeController.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.examination

import io.ebean.DB
import miscellaneous.datetime.DateTimeHandler
import miscellaneous.scala.DbApiHelper
import models.enrolment.ExamEnrolment
import models.user.Role
import org.joda.time.{DateTime, Seconds}
import play.api.mvc.*
import security.scala.Auth
import security.scala.Auth.{AuthenticatedAction, authorized}

import javax.inject.Inject
import scala.concurrent.ExecutionContext
import scala.util.Try

class TimeController @Inject() (
    dateTimeHandler: DateTimeHandler,
    val authenticated: AuthenticatedAction,
    val controllerComponents: ControllerComponents
)(implicit ec: ExecutionContext)
    extends BaseController
    with DbApiHelper:

  def getRemainingExamTime(hash: String): Action[AnyContent] =
    authenticated.andThen(authorized(Seq(Role.Name.STUDENT))) { request =>
      val user = request.attrs(Auth.ATTR_USER)

      DB.find(classOf[ExamEnrolment])
        .fetch("externalExam")
        .where()
        .disjunction()
        .eq("exam.hash", hash)
        .eq("externalExam.hash", hash)
        .endJunction()
        .eq("user.id", user.getId)
        .find match
        case None => NotFound
        case Some(enrolment) =>
          val start    = getStart(enrolment)
          val duration = getDuration(enrolment)
          val now      = currentTime(enrolment)
          val timeLeft = Seconds.secondsBetween(now, start.plusMinutes(duration))
          Ok(timeLeft.getSeconds.toString)
    }

  private def getStart(enrolment: ExamEnrolment): DateTime =
    Option(enrolment.getExaminationEventConfiguration) match
      case Some(config) => config.getExaminationEvent.getStart
      case None         => enrolment.getReservation.getStartAt

  private def currentTime(enrolment: ExamEnrolment): DateTime =
    Option(enrolment.getExaminationEventConfiguration) match
      case Some(_) => DateTime.now()
      case None    => dateTimeHandler.adjustDST(DateTime.now(), enrolment.getReservation)

  private def getDuration(enrolment: ExamEnrolment): Int =
    Option(enrolment.getExam) match
      case Some(exam) => exam.getDuration.intValue()
      case None =>
        Try(enrolment.getExternalExam.deserialize().getDuration.intValue()).getOrElse(0)
